examples/psp22-extension/runtime/psp22-extension-example.rs

```diff
@@ -0,0 +1,446 @@
+use codec::{
+    Decode,
+    Encode,
+    MaxEncodedLen,
+};
+use frame_support::{
+    dispatch::RawOrigin,
+    log::{
+        error,
+        trace,
+    },
+    pallet_prelude::*,
+    traits::fungibles::{
+        approvals::{
+            Inspect as AllowanceInspect,
+            Mutate as AllowanceMutate,
+        },
+        Inspect,
+        InspectMetadata,
+        Transfer,
+    },
+};
+use pallet_assets::{
+    self,
+    WeightInfo,
+};
+use pallet_contracts::chain_extension::{
+    ChainExtension,
+    Environment,
+    Ext,
+    InitState,
+    RetVal,
+    SysConfig,
+    UncheckedFrom,
+};
+use sp_runtime::{
+    traits::{
+        Saturating,
+        StaticLookup,
+        Zero,
+    },
+    DispatchError,
+};
+
+#[derive(Debug, PartialEq, Encode, Decode, MaxEncodedLen)]
+struct Psp22BalanceOfInput<AssetId, AccountId> {
+    asset_id: AssetId,
+    owner: AccountId,
+}
+
+#[derive(Debug, PartialEq, Encode, Decode, MaxEncodedLen)]
+struct Psp22AllowanceInput<AssetId, AccountId> {
+    asset_id: AssetId,
+    owner: AccountId,
+    spender: AccountId,
+}
+
+#[derive(Debug, PartialEq, Encode, Decode, MaxEncodedLen)]
+struct Psp22TransferInput<AssetId, AccountId, Balance> {
+    asset_id: AssetId,
+    to: AccountId,
+    value: Balance,
+}
+
+#[derive(Debug, PartialEq, Encode, Decode, MaxEncodedLen)]
+struct Psp22TransferFromInput<AssetId, AccountId, Balance> {
+    asset_id: AssetId,
+    from: AccountId,
+    to: AccountId,
+    value: Balance,
+}
+
+#[derive(Debug, PartialEq, Encode, Decode, MaxEncodedLen)]
+struct Psp22ApproveInput<AssetId, AccountId, Balance> {
+    asset_id: AssetId,
+    spender: AccountId,
+    value: Balance,
+}
+
+#[derive(Default)]
+pub struct Psp22Extension;
+
+fn convert_err(err_msg: &'static str) -> impl FnOnce(DispatchError) -> DispatchError {
+    move |err| {
+        trace!(
+            target: "runtime",
+            "PSP22 Transfer failed:{:?}",
+            err
+        );
+        DispatchError::Other(err_msg)
+    }
+}
+
+// We're using enums for function IDs because contrary to raw u16 it enables
+// exhaustive matching, which results in cleaner code.
+
+enum FuncId {
```

Comment:
```suggestion
/// We're using enums for function IDs because contrary to raw u16 it enables
/// exhaustive matching, which results in cleaner code.
enum FuncId {
```